adapter/internal/discovery/xds/server_utils.go

```diff
@@ -81,3 +85,110 @@ func updateVhostInternalMaps(uuid, name, version, vHost string, gwEnvs []string)
+		return nil, errors.New("Invalid version format")
+	}
+
+	if minorVersionConvErr != nil {
+		logger.LoggerXds.Errorf("API minor version should be an integer in API Version: %v", version, minorVersionConvErr)
+		return nil, errors.New("Invalid version format")
+	}
+
+	if len(versionComponents) == 2 {
+		return &SemVersion{
+			Version: version,
+			Major:   majorVersion,
+			Minor:   minorVersion,
+			Patch:   nil,
+		}, nil
+	}
+
+	patchVersion, patchVersionConvErr := strconv.Atoi(versionComponents[2])
+	if patchVersionConvErr != nil {
+		logger.LoggerXds.Errorf("API patch version should be an integer in API Version: %v", version, patchVersionConvErr)
+		return nil, errors.New("Invalid version format")
+	}
+	return &SemVersion{
+		Version: version,
+		Major:   majorVersion,
+		Minor:   minorVersion,
+		Patch:   &patchVersion,
+	}, nil
+}
+
+// GetMajorMinorVersionRangeRegex generates major and minor version compatible range regex for the given version
+func GetMajorMinorVersionRangeRegex(semVersion SemVersion) string {
+	version := semVersion.Version
+	majorVersion := strconv.Itoa(semVersion.Major)
+	minorVersion := strconv.Itoa(semVersion.Minor)
+	if semVersion.Patch == nil {
+		return "(v" + majorVersion + "(\\." + minorVersion + ")?|" + version + ")"
+	}
+	patchVersion := strconv.Itoa(*semVersion.Patch)
+	return "(v" + majorVersion + "(\\." + minorVersion + "(\\." + patchVersion + ")?)?|" + version + ")"
+}
+
+// GetMinorVersionRangeRegex generates minor version compatible range regex for the given version
+func GetMinorVersionRangeRegex(semVersion SemVersion) string {
+	version := semVersion.Version
+	if semVersion.Patch == nil {
+		return "(v" + version + "|" + version + ")"
+	}
+	majorVersion := strconv.Itoa(semVersion.Major)
+	minorVersion := strconv.Itoa(semVersion.Minor)
+	patchVersion := strconv.Itoa(*semVersion.Patch)
+	return "(v" + majorVersion + "\\." + minorVersion + "(\\." + patchVersion + ")?|" + version + ")"
+}
+
+// GetMajorVersionRange generates major version range for the given version
+func GetMajorVersionRange(semVersion SemVersion) string {
+	return "v" + strconv.Itoa(semVersion.Major)
+}
+
+// GetMinorVersionRange generates minor version range for the given version
+func GetMinorVersionRange(semVersion SemVersion) string {
+	return "v" + strconv.Itoa(semVersion.Major) + "." + strconv.Itoa(semVersion.Minor)
+}
+
+// CompareSemanticVersions compares two semantic versions and returns true
+// if `version` is greater or equal than `baseVersion`
+func CompareSemanticVersions(baseVersion, version SemVersion) bool {
+	if baseVersion.Major < version.Major {
+		return true
+	} else if baseVersion.Major > version.Major {
+		return false
+	} else {
+		if baseVersion.Minor < version.Minor {
+			return true
+		} else if baseVersion.Minor > version.Minor {
+			return false
+		} else {
+			if baseVersion.Patch != nil && version.Patch != nil {
+				if *baseVersion.Patch < *version.Patch {
```

Comment:
Can we use `baseVersion.Patch` in here as well ?

Reply:
Since the `Patch` attribute is declared as an optional attribute, we have to use with `*`